core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -0,0 +1,270 @@
+
+    private static final Pattern<JoinNode> PATTERN =
+            Patterns.join()
+                    .with(Patterns.Join.left().matching(tableScan().capturedAs(LEFT_TABLE_SCAN)))
+                    .with(Patterns.Join.right().matching(tableScan().capturedAs(RIGHT_TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushJoinIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        if (getJoinPushdownMode(context.getSession()) == JoinPushdownMode.OFF) {
+            return Result.empty();
+        }
+
+        if (joinNode.isCrossJoin()) {
+            return Result.empty();
+        }
+
+        TableScanNode left = captures.get(LEFT_TABLE_SCAN);
+        TableScanNode right = captures.get(RIGHT_TABLE_SCAN);
+
+        Expression effectiveFilter = getEffectiveFilter(joinNode);
+        FilterSplitResult filterSplitResult = splitFilter(effectiveFilter, joinNode.getLeftOutputSymbols(), joinNode.getRightOutputSymbols(), context);
+
+        if (!filterSplitResult.getRemainingFilter().equals(BooleanLiteral.TRUE_LITERAL)) {
+            // TODO add extra filter node above join
+            return Result.empty();
+        }
+
+        Map<String, ColumnHandle> leftAssignments = left.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Map<String, ColumnHandle> rightAssignments = right.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Optional<JoinApplicationResult<TableHandle>> joinApplicationResult = metadata.applyJoin(
+                context.getSession(),
+                getJoinType(joinNode),
+                left.getTable(),
+                right.getTable(),
+                filterSplitResult.getPushableConditions(),
+                // TODO we could send only subset of assignments here, those which are needed to resolve filterSplitResult.getPushableConditions
+                leftAssignments,
+                rightAssignments);
+
+        if (joinApplicationResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        TableHandle handle = joinApplicationResult.get().getTableHandle();
+
+        Map<ColumnHandle, ColumnHandle> columnHandlesMapping = joinApplicationResult.get().getColumnHandles();
+
+        TupleDomain<ColumnHandle> newEnforcedConstraints = TupleDomain.fromColumnDomains(
+                transformToNewAssignments(left.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
+                        leftDomains -> transformToNewAssignments(right.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
```

Comment:
i find this nested flatMap calls very clever. Maybe even a bit too clever for a regular reader like me.
however, there is no point to deal with TS-es with `TupleDomain.none` enforced constraints. We can simply bail out early if this is the case. Then this code doesn't have to deal with Optionals at all -- they now are known to be present.

Reply:
> i find this nested flatMap calls very clever

Thank you :)

Simplified following your advice.